File: sdks/java/src/main/java/co/marvincorporate/pay/sdk/MarvinPayException.java
package co.marvincorporate.pay.sdk;

/**
 * Thrown for any Marvin Pay SDK failure: a non-2xx HTTP response, a transport
 * error, or an interruption while polling.
 *
 * <p>For HTTP failures {@link #getHttpStatus()} carries the status code and
 * {@link #getBody()} the raw response body (which may be a Spring error object
 * or a {@code PaymentResult}-shaped payload with a non-2xx {@code status} and a
 * {@code message}). For transport/interruption failures {@link #getHttpStatus()}
 * is {@code 0} and {@link #getBody()} is {@code null}.
 */
public class MarvinPayException extends RuntimeException {

    private static final long serialVersionUID = 1L;

    private final int httpStatus;
    private final String body;

    /** Transport / client-side failure (no HTTP status). */
    public MarvinPayException(String message) {
        super(message);
        this.httpStatus = 0;
        this.body = null;
    }

    /** Transport / client-side failure with an underlying cause. */
    public MarvinPayException(String message, Throwable cause) {
        super(message, cause);
        this.httpStatus = 0;
        this.body = null;
    }

    /** Non-2xx HTTP response. */
    public MarvinPayException(int httpStatus, String message, String body) {
        super(message);
        this.httpStatus = httpStatus;
        this.body = body;
    }

    /** HTTP status code of the failing response, or {@code 0} for a non-HTTP failure. */
    public int getHttpStatus() {
        return httpStatus;
    }

    /** Raw response body of the failing response, or {@code null}. */
    public String getBody() {
        return body;
    }
}

File: sdks/php/src/MarvinPayException.php
<?php

declare(strict_types=1);

namespace MarvinPay;

/**
 * Thrown for any non-2xx HTTP response, transport/cURL failure, or client-side
 * validation error raised by {@see MarvinPayClient}.
 *
 * Carries the HTTP status (when there was a response), the human-readable
 * message (extracted from the response body's `message` field when present),
 * and the raw/decoded response body for inspection.
 */
class MarvinPayException extends \RuntimeException
{
    private ?int $httpStatus;

    /** @var mixed decoded array/object when JSON, raw string otherwise, or null */
    private $body;

    /**
     * @param mixed $body decoded body (array), raw string, or null
     */
    public function __construct(
        string $message,
        ?int $httpStatus = null,
        $body = null,
        ?\Throwable $previous = null
    ) {
        parent::__construct($message, $httpStatus ?? 0, $previous);
        $this->httpStatus = $httpStatus;
        $this->body = $body;
    }

    /** HTTP status code, or null for transport/client-side errors. */
    public function getHttpStatus(): ?int
    {
        return $this->httpStatus;
    }

    /**
     * The response body as decoded by the client (array for JSON, string for
     * non-JSON, or null when there was no body).
     *
     * @return mixed
     */
    public function getBody()
    {
        return $this->body;
    }
}
